From 886a1161946635a6b7b198eded0c1793e12e7700 Mon Sep 17 00:00:00 2001
From: Vinod Paidimarry <vinodkiran@outlook.in>
Date: Fri, 28 Feb 2025 08:39:12 +0530
Subject: [PATCH 1/2] adding redaction rules to post processing

---
 .../ui-component/extended/PostProcessing.jsx  | 371 +++++++++++++++---
 1 file changed, 307 insertions(+), 64 deletions(-)

diff --git a/packages/ui/src/ui-component/extended/PostProcessing.jsx b/packages/ui/src/ui-component/extended/PostProcessing.jsx
index fd56a3eb683..6180405f8c6 100644
--- a/packages/ui/src/ui-component/extended/PostProcessing.jsx
+++ b/packages/ui/src/ui-component/extended/PostProcessing.jsx
@@ -4,13 +4,14 @@ import PropTypes from 'prop-types'
 import { useSelector } from 'react-redux'
 
 // material-ui
-import { IconButton, Button, Box, Typography } from '@mui/material'
-import { IconArrowsMaximize, IconBulb, IconX } from '@tabler/icons-react'
+import { IconButton, Button, Box, Typography, Tabs, Tab, Stack, OutlinedInput, Divider, Chip, Breadcrumbs } from '@mui/material'
+import { IconArrowRight, IconArrowsMaximize, IconBulb, IconX } from '@tabler/icons-react'
 import { useTheme } from '@mui/material/styles'
 
 // Project import
 import { StyledButton } from '@/ui-component/button/StyledButton'
 import { SwitchInput } from '@/ui-component/switch/Switch'
+import { TabPanel } from '@/ui-component/tabs/TabPanel'
 import { CodeEditor } from '@/ui-component/editor/CodeEditor'
 import ExpandTextDialog from '@/ui-component/dialog/ExpandTextDialog'
 
@@ -20,8 +21,47 @@ import useNotifier from '@/utils/useNotifier'
 
 // API
 import chatflowsApi from '@/api/chatflows'
+import { Dropdown } from '@/ui-component/dropdown/Dropdown'
 
 const sampleFunction = `return $flow.rawOutput + " This is a post processed response!";`
+const redactionTypes = [
+    {
+        label: 'Remove matching text',
+        name: 'remove'
+    },
+    {
+        label: 'Replace with custom text',
+        name: 'replace'
+    },
+    {
+        label: 'Replace with ***',
+        name: 'asterisk'
+    }
+]
+
+function HorizontalStepper() {
+    const breadcrumbs = [
+        <Typography key='1' style={{ color: '#b22c4d' }}>
+            Generate LLM output
+        </Typography>,
+        <Typography key='2' style={{ color: '#b22c4d' }}>
+            Apply Redaction Rules
+        </Typography>,
+        <Typography key='3' style={{ color: '#b22c4d' }}>
+            Execute Custom JS Function
+        </Typography>,
+        <Typography key='4' style={{ color: '#b22c4d' }}>
+            Show Response
+        </Typography>
+    ]
+    return (
+        <Stack spacing={2} style={{ alignItems: 'center' }}>
+            <Breadcrumbs separator={<IconArrowRight fontSize='small' />} aria-label='breadcrumb'>
+                {breadcrumbs}
+            </Breadcrumbs>
+        </Stack>
+    )
+}
 
 const PostProcessing = ({ dialogProps }) => {
     const dispatch = useDispatch()
@@ -39,6 +79,38 @@ const PostProcessing = ({ dialogProps }) => {
     const [showExpandDialog, setShowExpandDialog] = useState(false)
     const [expandDialogProps, setExpandDialogProps] = useState({})
 
+    const [activeTabValue, setActiveTabValue] = useState(0)
+
+    const [replacementText, setReplacementText] = useState('')
+    const [redactionPhrase, setRedactionPhrase] = useState('')
+    const [isRegexp, setRegexp] = useState(false)
+    const [redactionType, setRedactionType] = useState('asterisk')
+    const [redactionRules, setRedactionRules] = useState([])
+
+    const addRedactionRule = () => {
+        const newRule = {
+            phrase: redactionPhrase,
+            isRegexp: isRegexp,
+            type: redactionType,
+            replacement: redactionType === 'replace' ? replacementText : ''
+        }
+        setRedactionRules([...redactionRules, newRule])
+        setRedactionPhrase('')
+        setRedactionType('asterisk')
+        setReplacementText('')
+        setRegexp(false)
+    }
+
+    const removeRedactionRule = (index) => {
+        const rows = [...redactionRules]
+        rows.splice(index, 1)
+        setRedactionRules(rows)
+    }
+
+    const handleTabChange = (event, newValue) => {
+        setActiveTabValue(newValue)
+    }
+
     const handleChange = (value) => {
         setPostProcessingEnabled(value)
     }
@@ -119,87 +191,258 @@ const PostProcessing = ({ dialogProps }) => {
             }
         }
 
-        return () => {}
+        return () => {
+            setRedactionRules([...redactionRules, newRule])
+            setRedactionPhrase('')
+            setRedactionType('asterisk')
+            setReplacementText('')
+            setRegexp(false)
+        }
     }, [dialogProps])
 
     return (
         <>
-            <Box sx={{ width: '100%', display: 'flex', alignItems: 'center', justifyContent: 'space-between', mb: 2 }}>
-                <SwitchInput label='Enable Post Processing' onChange={handleChange} value={postProcessingEnabled} />
-            </Box>
-            <Box sx={{ width: '100%', display: 'flex', flexDirection: 'column', alignItems: 'flex-start', gap: 1 }}>
-                <Box sx={{ width: '100%', display: 'flex', alignItems: 'center' }}>
-                    <Typography>JS Function</Typography>
-                    <Button
-                        sx={{ ml: 2 }}
-                        variant='outlined'
-                        onClick={() => {
-                            setPostProcessingFunction(sampleFunction)
-                        }}
-                    >
-                        See Example
-                    </Button>
-                    <div style={{ flex: 1 }} />
-                    <IconButton
-                        size='small'
-                        sx={{
-                            height: 25,
-                            width: 25
-                        }}
-                        title='Expand'
-                        color='primary'
-                        onClick={() => onExpandDialogClicked(postProcessingFunction)}
-                    >
-                        <IconArrowsMaximize />
-                    </IconButton>
-                </Box>
-
-                <div
-                    style={{
-                        marginTop: '10px',
-                        border: '1px solid',
-                        borderColor: theme.palette.grey['300'],
-                        borderRadius: '6px',
-                        height: '200px',
-                        width: '100%'
-                    }}
-                >
-                    <CodeEditor
-                        value={postProcessingFunction}
-                        height='200px'
-                        theme={customization.isDarkMode ? 'dark' : 'light'}
-                        lang={'js'}
-                        placeholder={sampleFunction}
-                        onValueChange={(code) => setPostProcessingFunction(code)}
-                        basicSetup={{ highlightActiveLine: false, highlightActiveLineGutter: false }}
-                    />
-                </div>
-            </Box>
             <div
                 style={{
                     display: 'flex',
+                    alignItems: 'center',
                     flexDirection: 'column',
                     borderRadius: 10,
-                    background: '#d8f3dc',
+                    background: '#f3bdcc',
                     padding: 10,
                     marginTop: 10
                 }}
             >
+                <span style={{ marginLeft: 10, fontWeight: 500, color: '#000000' }}>
+                    <HorizontalStepper />
+                </span>
+            </div>
+
+            <Box sx={{ width: '100%', display: 'flex', alignItems: 'center', justifyContent: 'space-between', mb: 2 }}>
+                <SwitchInput label='Enable Post Processing' onChange={handleChange} value={postProcessingEnabled} />
+            </Box>
+            <Tabs
+                value={activeTabValue}
+                onChange={handleTabChange}
+                textColor='primary'
+                aria-label='tabs'
+                style={{ marginBottom: 10, background: '#f5f5f5' }}
+            >
+                <Tab value={0} label='Redaction Rules' />
+                <Tab value={1} label='Custom JS Function'></Tab>
+            </Tabs>
+            <TabPanel value={activeTabValue} index={0}>
+                <Stack direction='column' divider={<Divider sx={{ my: 0.5, borderColor: theme.palette.grey[900] + 25 }} />} spacing={2}>
+                    <Stack direction='column' spacing={2} sx={{ width: '100%' }}>
+                        <Stack direction='row' spacing={1} style={{ width: '100%' }}>
+                            <Stack direction='column' spacing={2} style={{ width: '90%' }}>
+                                <Typography>Phrase or Pattern</Typography>
+                                <div style={{ display: 'flex', width: '100%' }}>
+                                    <Box sx={{ width: '100%', mb: 1 }}>
+                                        <OutlinedInput
+                                            sx={{ width: '100%' }}
+                                            type='text'
+                                            onChange={(e) => setRedactionPhrase(e.target.value)}
+                                            size='small'
+                                            value={redactionPhrase}
+                                            name='redactionPhrase'
+                                            placeholder='Enter a phrase or regexp to redact'
+                                        />
+                                    </Box>
+                                </div>
+                            </Stack>
+                            <Stack direction='column' spacing={2} style={{ width: '10%' }}>
+                                <Typography>Is RegExp?</Typography>
+                                <div style={{ display: 'flex', width: '100%' }}>
+                                    <Box sx={{ width: '100%', mb: 1 }}>
+                                        <SwitchInput label={''} onChange={setRegexp} value={isRegexp} />
+                                    </Box>
+                                </div>
+                            </Stack>
+                        </Stack>
+                        <Stack direction='row' spacing={1} style={{ width: '100%' }}>
+                            <Stack direction='column' spacing={1} style={{ width: redactionType === 'replace' ? '50%' : '100%' }}>
+                                <Typography>Action</Typography>
+                                <Dropdown
+                                    key={redactionType}
+                                    name={redactionType}
+                                    options={redactionTypes}
+                                    onSelect={(newValue) => setRedactionType(newValue)}
+                                    value={redactionType ?? 'choose an option'}
+                                />
+                            </Stack>
+                            {redactionType === 'replace' && (
+                                <Stack direction='column' spacing={1} style={{ width: '50%' }}>
+                                    <Typography>Replacement Text</Typography>
+                                    <OutlinedInput
+                                        sx={{ width: '100%' }}
+                                        type='text'
+                                        onChange={(e) => setReplacementText(e.target.value)}
+                                        size='small'
+                                        value={replacementText}
+                                        name='replacementText'
+                                        placeholder='Enter replacement text'
+                                    />
+                                </Stack>
+                            )}
+                        </Stack>
+                        <Stack direction='column' spacing={1}>
+                            <StyledButton
+                                style={{ marginBottom: 5, marginTop: 5 }}
+                                variant='contained'
+                                disabled={!redactionPhrase || !redactionType || (redactionType === 'replace' && !replacementText)}
+                                onClick={addRedactionRule}
+                            >
+                                Add Redaction rule
+                            </StyledButton>
+                        </Stack>
+                    </Stack>
+                    <Stack direction='column' spacing={2} sx={{ width: '100%' }}>
+                        <Box sx={{ width: '100%', display: 'flex', alignItems: 'left', margin: 2 }}>
+                            <Typography variant='h3'>Active Redaction Rules</Typography>
+                        </Box>
+                        <Stack direction='column' spacing={1} style={{ width: '100%' }}>
+                            {redactionRules.map((rule, index) => (
+                                <Stack
+                                    key={index}
+                                    direction='row'
+                                    spacing={1}
+                                    style={{
+                                        padding: 2,
+                                        margin: 2,
+                                        width: '100%',
+                                        alignItems: 'center',
+                                        background: '#f6f5f5',
+                                        justifyContent: 'space-between'
+                                    }}
+                                >
+                                    <Box
+                                        sx={{
+                                            width: '100%',
+                                            display: 'flex',
+                                            alignItems: 'left',
+                                            verticalAlign: 'middle'
+                                        }}
+                                    >
+                                        {rule.isRegexp ? (
+                                            <Chip
+                                                size='small'
+                                                variant='filled'
+                                                style={{ background: 'black', color: 'white', marginRight: 5 }}
+                                                label='Regexp'
+                                            />
+                                        ) : (
+                                            <Chip
+                                                size='small'
+                                                variant='filled'
+                                                style={{ background: 'black', color: 'white', marginRight: 5 }}
+                                                label='Text'
+                                            />
+                                        )}
+                                        {rule.type === 'remove' && (
+                                            <Typography
+                                                style={{ color: 'darkblue' }}
+                                                variant='subtitle1'
+                                            >{`Remove: ${rule.phrase}`}</Typography>
+                                        )}
+                                        {rule.type === 'replace' && (
+                                            <Typography
+                                                style={{ color: 'darkblue' }}
+                                                variant='subtitle1'
+                                            >{`Replace: ${rule.phrase} with ${rule.replacement}`}</Typography>
+                                        )}
+                                        {rule.type === 'asterisk' && (
+                                            <Typography
+                                                style={{ color: 'darkblue' }}
+                                                variant='subtitle1'
+                                            >{`Replace: ${rule.phrase} with ***`}</Typography>
+                                        )}
+                                    </Box>
+                                    <Button onClick={() => removeRedactionRule(index)}>
+                                        <IconX />
+                                    </Button>
+                                </Stack>
+                            ))}
+                            {redactionRules.length === 0 && <Typography>No active redaction rules</Typography>}
+                        </Stack>
+                    </Stack>
+                </Stack>
+            </TabPanel>
+            <TabPanel value={activeTabValue} index={1}>
+                <Box sx={{ width: '100%', display: 'flex', flexDirection: 'column', alignItems: 'flex-start', gap: 1 }}>
+                    <Box sx={{ width: '100%', display: 'flex', alignItems: 'center' }}>
+                        <Typography>JS Function</Typography>
+                        <Button
+                            sx={{ ml: 2 }}
+                            variant='outlined'
+                            onClick={() => {
+                                setPostProcessingFunction(sampleFunction)
+                            }}
+                        >
+                            See Example
+                        </Button>
+                        <div style={{ flex: 1 }} />
+                        <IconButton
+                            size='small'
+                            sx={{
+                                height: 25,
+                                width: 25
+                            }}
+                            title='Expand'
+                            color='primary'
+                            onClick={() => onExpandDialogClicked(postProcessingFunction)}
+                        >
+                            <IconArrowsMaximize />
+                        </IconButton>
+                    </Box>
+
+                    <div
+                        style={{
+                            marginTop: '10px',
+                            border: '1px solid',
+                            borderColor: theme.palette.grey['300'],
+                            borderRadius: '6px',
+                            height: '200px',
+                            width: '100%'
+                        }}
+                    >
+                        <CodeEditor
+                            value={postProcessingFunction}
+                            height='200px'
+                            theme={customization.isDarkMode ? 'dark' : 'light'}
+                            lang={'js'}
+                            placeholder={sampleFunction}
+                            onValueChange={(code) => setPostProcessingFunction(code)}
+                            basicSetup={{ highlightActiveLine: false, highlightActiveLineGutter: false }}
+                        />
+                    </div>
+                </Box>
                 <div
                     style={{
                         display: 'flex',
-                        flexDirection: 'row',
-                        alignItems: 'center',
-                        paddingTop: 10
+                        flexDirection: 'column',
+                        borderRadius: 10,
+                        background: '#d8f3dc',
+                        padding: 10,
+                        marginTop: 10
                     }}
                 >
-                    <IconBulb size={30} color='#2d6a4f' />
-                    <span style={{ color: '#2d6a4f', marginLeft: 10, fontWeight: 500 }}>
-                        The following variables are available to use in the custom function:{' '}
-                        <pre>$flow.rawOutput, $flow.input, $flow.chatflowId, $flow.sessionId, $flow.chatId</pre>
-                    </span>
+                    <div
+                        style={{
+                            display: 'flex',
+                            flexDirection: 'row',
+                            alignItems: 'center',
+                            paddingTop: 10
+                        }}
+                    >
+                        <IconBulb size={30} color='#2d6a4f' />
+                        <span style={{ color: '#2d6a4f', marginLeft: 10, fontWeight: 500 }}>
+                            The following variables are available to use in the custom function:{' '}
+                            <pre>$flow.rawOutput, $flow.input, $flow.chatflowId, $flow.sessionId, $flow.chatId</pre>
+                        </span>
+                    </div>
                 </div>
-            </div>
+            </TabPanel>
             <StyledButton
                 style={{ marginBottom: 10, marginTop: 10 }}
                 variant='contained'

From 047d8ad2cf08ac6007571aed8d4203d7d5821cd4 Mon Sep 17 00:00:00 2001
From: Vinod Paidimarry <vinodkiran@outlook.in>
Date: Fri, 28 Feb 2025 17:45:06 +0530
Subject: [PATCH 2/2] Post Processing -> Redaction Rules

---
 packages/server/src/Interface.ts              | 12 +++
 packages/server/src/utils/buildChatflow.ts    | 74 ++++++++++++++-----
 .../ui-component/extended/PostProcessing.jsx  | 21 ++++--
 3 files changed, 81 insertions(+), 26 deletions(-)

diff --git a/packages/server/src/Interface.ts b/packages/server/src/Interface.ts
index fbf4ea1299e..40589ff50e0 100644
--- a/packages/server/src/Interface.ts
+++ b/packages/server/src/Interface.ts
@@ -354,3 +354,15 @@ export interface IVariableOverride {
 
 // DocumentStore related
 export * from './Interface.DocumentStore'
+
+export enum RedactionRuleType {
+    REMOVE = 'remove',
+    REPLACE = 'replace',
+    ASTERISK = 'asterisk'
+}
+export interface RedactionRule {
+    phrase: string
+    isRegexp: boolean
+    type: RedactionRuleType
+    replacement: string
+}
diff --git a/packages/server/src/utils/buildChatflow.ts b/packages/server/src/utils/buildChatflow.ts
index f73f6044398..86d5dbd73c6 100644
--- a/packages/server/src/utils/buildChatflow.ts
+++ b/packages/server/src/utils/buildChatflow.ts
@@ -32,7 +32,9 @@ import {
     IVariable,
     INodeOverrides,
     IVariableOverride,
-    MODE
+    MODE,
+    RedactionRule,
+    RedactionRuleType
 } from '../Interface'
 import { InternalFlowiseError } from '../errors/internalFlowiseError'
 import { databaseEntities } from '.'
@@ -653,26 +655,62 @@ export const executeFlow = async ({
             resultText = result.text
             /* Check for post-processing settings */
             if (chatflowConfig?.postProcessing?.enabled === true) {
+                let moderatedResponse = resultText
                 try {
-                    const postProcessingFunction = JSON.parse(chatflowConfig?.postProcessing?.customFunction)
-                    const nodeInstanceFilePath = componentNodes['customFunction'].filePath as string
-                    const nodeModule = await import(nodeInstanceFilePath)
-                    const nodeData = {
-                        inputs: { javascriptFunction: postProcessingFunction },
-                        outputs: { output: 'output' }
+                    //first run redaction rules
+                    if (chatflowConfig.postProcessing.redactionRules) {
+                        const redactionRules: RedactionRule[] = JSON.parse(chatflowConfig.postProcessing.redactionRules)
+                        for (const rule of redactionRules) {
+                            switch (rule.type) {
+                                case RedactionRuleType.REMOVE:
+                                    if (rule.isRegexp) {
+                                        const regex = new RegExp(rule.phrase, 'g')
+                                        moderatedResponse = moderatedResponse.replace(regex, '')
+                                    } else {
+                                        moderatedResponse = moderatedResponse.replace(rule.phrase, '')
+                                    }
+                                    break
+                                case RedactionRuleType.REPLACE:
+                                    if (rule.isRegexp) {
+                                        const regex = new RegExp(rule.phrase, 'g')
+                                        moderatedResponse = moderatedResponse.replace(regex, rule.replacement)
+                                    } else {
+                                        moderatedResponse = moderatedResponse.replace(rule.phrase, rule.replacement)
+                                    }
+                                    break
+                                case RedactionRuleType.ASTERISK:
+                                    if (rule.isRegexp) {
+                                        const regex = new RegExp(rule.phrase, 'g')
+                                        moderatedResponse = moderatedResponse.replace(regex, '***')
+                                    } else {
+                                        moderatedResponse = moderatedResponse.replace(rule.phrase, '***')
+                                    }
+                                    break
+                            }
+                        }
                     }
-                    const options: ICommonObject = {
-                        chatflowid: chatflow.id,
-                        sessionId,
-                        chatId,
-                        input: question,
-                        rawOutput: resultText,
-                        appDataSource,
-                        databaseEntities,
-                        logger
+                    //second - apply custom function
+                    if (chatflowConfig?.postProcessing?.customFunction) {
+                        const postProcessingFunction = JSON.parse(chatflowConfig?.postProcessing?.customFunction)
+                        const nodeInstanceFilePath = componentNodes['customFunction'].filePath as string
+                        const nodeModule = await import(nodeInstanceFilePath)
+                        const nodeData = {
+                            inputs: { javascriptFunction: postProcessingFunction },
+                            outputs: { output: 'output' }
+                        }
+                        const options: ICommonObject = {
+                            chatflowid: chatflow.id,
+                            sessionId,
+                            chatId,
+                            input: question,
+                            rawOutput: moderatedResponse,
+                            appDataSource,
+                            databaseEntities,
+                            logger
+                        }
+                        const customFuncNodeInstance = new nodeModule.nodeClass()
+                        moderatedResponse = await customFuncNodeInstance.init(nodeData, question, options)
                     }
-                    const customFuncNodeInstance = new nodeModule.nodeClass()
-                    let moderatedResponse = await customFuncNodeInstance.init(nodeData, question, options)
                     result.text = moderatedResponse
                     resultText = result.text
                 } catch (e) {
diff --git a/packages/ui/src/ui-component/extended/PostProcessing.jsx b/packages/ui/src/ui-component/extended/PostProcessing.jsx
index 6180405f8c6..62b144c9c3e 100644
--- a/packages/ui/src/ui-component/extended/PostProcessing.jsx
+++ b/packages/ui/src/ui-component/extended/PostProcessing.jsx
@@ -55,7 +55,8 @@ function HorizontalStepper() {
         </Typography>
     ]
     return (
-        <Stack spacing={2} style={{ alignItems: 'center' }}>
+        <Stack spacing={2} style={{ alignItems: 'center', color: 'black' }}>
+            <Typography variant={'h5'}>When enabled, the order of Execution</Typography>
             <Breadcrumbs separator={<IconArrowRight fontSize='small' />} aria-label='breadcrumb'>
                 {breadcrumbs}
             </Breadcrumbs>
@@ -138,6 +139,7 @@ const PostProcessing = ({ dialogProps }) => {
             let value = {
                 postProcessing: {
                     enabled: postProcessingEnabled,
+                    redactionRules: JSON.stringify(redactionRules),
                     customFunction: JSON.stringify(postProcessingFunction)
                 }
             }
@@ -188,11 +190,14 @@ const PostProcessing = ({ dialogProps }) => {
                 if (chatbotConfig.postProcessing.customFunction) {
                     setPostProcessingFunction(JSON.parse(chatbotConfig.postProcessing.customFunction))
                 }
+                if (chatbotConfig.postProcessing.redactionRules) {
+                    setRedactionRules(JSON.parse(chatbotConfig.postProcessing.redactionRules))
+                }
             }
         }
 
         return () => {
-            setRedactionRules([...redactionRules, newRule])
+            setRedactionRules([])
             setRedactionPhrase('')
             setRedactionType('asterisk')
             setReplacementText('')
@@ -202,25 +207,25 @@ const PostProcessing = ({ dialogProps }) => {
 
     return (
         <>
+            <Box sx={{ width: '100%', display: 'flex', alignItems: 'center', justifyContent: 'space-between', mb: 2 }}>
+                <SwitchInput label='Enable Post Processing' onChange={handleChange} value={postProcessingEnabled} />
+            </Box>
             <div
                 style={{
                     display: 'flex',
                     alignItems: 'center',
                     flexDirection: 'column',
                     borderRadius: 10,
-                    background: '#f3bdcc',
+                    background: '#fceaf0',
                     padding: 10,
-                    marginTop: 10
+                    marginTop: 5,
+                    marginBottom: 5
                 }}
             >
                 <span style={{ marginLeft: 10, fontWeight: 500, color: '#000000' }}>
                     <HorizontalStepper />
                 </span>
             </div>
-
-            <Box sx={{ width: '100%', display: 'flex', alignItems: 'center', justifyContent: 'space-between', mb: 2 }}>
-                <SwitchInput label='Enable Post Processing' onChange={handleChange} value={postProcessingEnabled} />
-            </Box>
             <Tabs
                 value={activeTabValue}
                 onChange={handleTabChange}
